src/Microsoft.AspNet.OData.Shared/Formatter/ODataOutputFormatterHelper.cs

```diff
@@ -349,5 +348,52 @@ private static bool IsOperationPath(ODataPath path)
 
             return false;
         }
+
+        private static Microsoft.OData.UriParser.ODataPath ConvertPath(ODataPath path)
+        {
+            if (path == null)
+            {
+                return null;
+            }
+
+            if (IsOperationPath(path))
+            {
+                var lastSegment = path.Segments.Last();
+                OperationSegment operation = lastSegment as OperationSegment;
+                if (operation != null && operation.EntitySet != null)
+                {
+                    return GeneratePath(operation.EntitySet);
+                }
+
+                OperationImportSegment operationImport = lastSegment as OperationImportSegment;
+                if (operationImport != null && operationImport.EntitySet != null)
+                {
+                    return GeneratePath(operationImport.EntitySet);
+                }
+
+                return null;
+            }
+
+            return path.Path;
+        }
+
+        private static Microsoft.OData.UriParser.ODataPath GeneratePath(IEdmEntitySetBase entitySetBase)
+        {
+            switch (entitySetBase.NavigationSourceKind())
+            {
+                case EdmNavigationSourceKind.EntitySet:
+                    return new Microsoft.OData.UriParser.ODataPath(new EntitySetSegment((IEdmEntitySet)entitySetBase));
+
+                case EdmNavigationSourceKind.Singleton:
+                    return new Microsoft.OData.UriParser.ODataPath(new SingletonSegment((IEdmSingleton)entitySetBase));
+
+                case EdmNavigationSourceKind.ContainedEntitySet:
+                    // TODO: for containment, do we have the scenario to return a containment in an operation?
```

Comment:
Yes; this should be supported.  This may seem like an edge case, but for services like Microsoft Graph that organize different sources of data through a singleton navigation property it is actually the common scenario.

Reply:
I add codes to process this case. Please take a look

---
In reply to: [385351278](https://github.com/OData/WebApi/pull/2060#discussion_r385351278) [](ancestors = 385351278)